skyportal/facility_apis/swift.py

```diff
@@ -150,6 +244,131 @@ def _build_payload(self, request):
+    """
+
+    from ..models import (
+        Comment,
+        DBSession,
+        FollowupRequest,
+        Group,
+    )
+    from swifttools.swift_too import Data
+
+    Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
+    session = Session()
+
+    try:
+        req = session.scalars(
+            sa.select(FollowupRequest).where(FollowupRequest.id == request_id)
+        ).first()
+
+        group_ids = [g.id for g in req.requester.accessible_groups]
+        groups = session.scalars(
+            Group.select(req.requester).where(Group.id.in_(group_ids))
+        ).all()
+
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            obsids = list({row.obsid for row in oq})
+            for obsid in obsids:
+                data = Data()
+                data.obsid = obsid
+                data.xrt = req.payload.get("XRT", False)
+                data.uvot = req.payload.get("UVOT", False)
+                data.bat = req.payload.get("BAT", False)
+                data.outdir = tmpdirname
+
+                if data.submit():
+                    data.outdir = os.path.expanduser(data.outdir)
+                    data.outdir = os.path.expandvars(data.outdir)
+                    data.outdir = os.path.abspath(data.outdir)
+
+                    # Index any existing files
+                    for i in range(len(data.entries)):
+                        fullfilepath = os.path.join(
+                            data.outdir, data.entries[i].path, data.entries[i].filename
+                        )
+                        if os.path.exists(fullfilepath):
+                            data.entries[i].localpath = fullfilepath
+                    topdir = os.path.join(data.outdir, str(obsid))
+                    if not os.path.isdir(topdir):
+                        os.makedirs(topdir)
+
+                    for dfile in data.entries:
+                        if not dfile.download(outdir=data.outdir):
+                            raise ValueError(f"Error downloading {dfile.filename}")
+                    filename = os.path.join(tmpdirname, f'{obsid}.tar.gz')
+                    with tarfile.open(filename, "w:gz") as tar:
+                        tar.add(topdir, arcname=os.path.basename(topdir))
+
+                    attachment_name = filename.split("/")[-1]
+                    with open(filename, 'rb') as f:
+                        attachment_bytes = base64.b64encode(f.read())
+                    comment = Comment(
+                        text=f'Swift Data: {obsid}',
+                        obj_id=req.obj.id,
+                        attachment_bytes=attachment_bytes,
+                        attachment_name=attachment_name,
+                        author=req.requester,
+                        groups=groups,
+                        bot=False,
+                    )
+                    session.add(comment)
+        req.status = 'Result posted as comment'
+        session.commit()
```

Comment:
I'm a little worried that we post things without any permissions checks.

Reply:
I guess the theory is that you had to pass all of those to use this handler (i.e. you make a request through the followup request handler, which calls this stuff). But it is a bit niddly.